ledger-service/http-json/src/main/scala/com/digitalasset/http/query/ValuePredicate.scala

```diff
@@ -0,0 +1,250 @@
+    fromJsObject(it, iface.TypeCon(iface.TypeConName(lfIdentifier(typ)), ImmArraySeq.empty), defs)
+
+  def fromJsObject(it: Map[String, JsValue], typ: iface.Type, defs: TypeLookup): ValuePredicate = {
+    type Result = ValuePredicate
+
+    def fromValue(it: JsValue, typ: iface.Type): Result =
+      (typ, it).match2 {
+        case p @ iface.TypePrim(_, _) => { case _ => fromPrim(it, p) }
+        case tc @ iface.TypeCon(iface.TypeConName(id), _) => {
+          case _ =>
+            val ddt = defs(id).getOrElse(predicateParseError(s"Type $id not found"))
+            fromCon(it, id, tc instantiate ddt)
+        }
+        case iface.TypeNumeric(scale) => {
+          case JsString(q) =>
+            val nq = Numeric checkWithinBoundsAndRound (scale, BigDecimal(q)) fold (predicateParseError, identity)
+            Literal { case V.ValueNumeric(v) if nq == (v setScale scale) => }
+          case JsNumber(q) =>
+            val nq = Numeric checkWithinBoundsAndRound (scale, q) fold (predicateParseError, identity)
+            Literal { case V.ValueNumeric(v) if nq == (v setScale scale) => }
+        }
+        case iface.TypeVar(_) => predicateParseError("no vars allowed!")
+      }(fallback = illTypedQuery(it, typ))
+
+    def fromCon(it: JsValue, id: Ref.Identifier, typ: iface.DataType.FWT): Result =
+      (typ, it).match2 {
+        case rec @ iface.Record(_) => {
+          case JsObject(fields) =>
+            fromRecord(fields, id, rec)
+        }
+        case iface.Variant(fieldTyps) => {
+          case JsObject(fields) =>
+            fromVariant(fields, id, fieldTyps)
+        }
+        case e @ iface.Enum(_) => {
+          case JsString(s) => fromEnum(s, id, e)
+        }
+      }(fallback = ???)
+
+    def fromRecord(
+        fields: Map[String, JsValue],
+        id: Ref.Identifier,
+        typ: iface.Record.FWT): Result = {
+      val iface.Record(fieldTyps) = typ
+      val invalidKeys = fields.keySet diff fieldTyps.iterator.map(_._1).toSet
+      if (invalidKeys.nonEmpty)
+        predicateParseError(s"$id does not have fields $invalidKeys")
+      RecordSubset(fieldTyps map {
+        case (fName, fTy) =>
+          fields get fName map (fSpec => (fName, fromValue(fSpec, fTy)))
+      })
+    }
+
+    def fromVariant(
+        fields: Map[String, JsValue],
+        id: Ref.Identifier,
+        fieldTyps: ImmArraySeq[(Ref.Name, Ty)]): Result = fields.toSeq match {
+      case Seq((k, v)) =>
+        val name = Ref.Name.assertFromString(k)
+        val field: Option[(Ref.Name, Ty)] = fieldTyps.find(_._1 == name)
+        val fieldP: Option[(Ref.Name, ValuePredicate)] = field.map {
+          case (n, t) => (n, fromValue(v, t))
+        }
+        fieldP.fold(
+          predicateParseError(
+            s"Cannot locate Variant's (datacon, type) field, id: $id, name: $name")
+        )(VariantMatch)
+
+      case _ => predicateParseError(s"Variant must have exactly 1 field, got: $fields, id: $id")
```

Comment:
You cannot really inline them, you still need the parent field, which is a `DataCon` -> `Record`
where DataCon is data constructor, the name of the variant and record is a nested record which can be empty. If you inline the record, where are you going to put the data constructor string? This would require changing the ADT.

Reply:
> What do you think about "inline" the fields of the child record as a shortcut?

With the caveat that users would have to be aware of exceptions, exceptions to those exceptions, and exceptions to those exceptions to those exceptions (I stopped thinking of examples at that point).

> Especially considering that the name of the variant field is not directly "visible" in the DAML code. What field name should users use?

The name is visible in DAML code; it's the data constructor.

Reply:
Okay, thanks.